fs/btrfs/zstd.c

```diff
@@ -50,10 +51,33 @@ static void zstd_free_workspace(struct list_head *ws)
 	kfree(workspace);
 }
 
-static struct list_head *zstd_alloc_workspace(void)
+static int zstd_set_level(struct list_head *ws, unsigned level)
+{
+	struct workspace *workspace = list_entry(ws, struct workspace, list);
+	ZSTD_parameters params;
+	int size;
+
+	if (level > ZSTD_BTRFS_MAX_LEVEL)
+		level = ZSTD_BTRFS_MAX_LEVEL;
+
+	if (level == 0)
+		level = ZSTD_BTRFS_DEFAULT_LEVEL;
+
+	params = ZSTD_getParams(level, ZSTD_BTRFS_MAX_INPUT, 0);
+	size = max_t(size_t,
+			ZSTD_CStreamWorkspaceBound(params.cParams),
+			ZSTD_DStreamWorkspaceBound(ZSTD_BTRFS_MAX_INPUT));
+	if (size > workspace->size) {
+		return -1;
+	}
```

Comment:
I think we could. The size for `buf` doesn't change, so we can reuse that too. I think we'll just need to be careful with [memalloc_nofs_save()](https://github.com/torvalds/linux/blob/master/fs/btrfs/compression.c#L902), since we call `kvmalloc()`.